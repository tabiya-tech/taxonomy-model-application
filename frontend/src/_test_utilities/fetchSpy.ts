export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
