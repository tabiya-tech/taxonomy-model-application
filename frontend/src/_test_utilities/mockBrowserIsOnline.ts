import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
