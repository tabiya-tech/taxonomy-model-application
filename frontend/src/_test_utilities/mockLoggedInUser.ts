import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
