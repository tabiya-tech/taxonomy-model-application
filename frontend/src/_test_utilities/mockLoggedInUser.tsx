import {
  AuthContextValue,
  TabiyaUser,
  authContextDefaultValue,
  AuthContext,
  AuthProviderProps,
} from "src/auth/AuthProvider";
import AuthAPISpecs from "api-specifications/auth";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUsers.Anonymous;

  const value = {
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: AuthAPISpecs.Enums.TabiyaRoles) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues,
  };

  // @ts-ignore
  jest.spyOn(require("src/auth/AuthProvider"), "AuthProvider").mockImplementation((props: AuthProviderProps) => {
    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>;
  });
}

export const TestUsers = {
  ModelManager: {
    username: "ModelManagerUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER],
  },
  RegisteredUser: {
    username: "RegisteredUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER],
  },
  Anonymous: {
    username: "AnonymousUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS],
  },
};
