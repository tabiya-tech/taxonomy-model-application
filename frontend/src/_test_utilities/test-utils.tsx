// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
