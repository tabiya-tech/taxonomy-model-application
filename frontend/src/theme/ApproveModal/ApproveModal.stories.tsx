import ApproveModal from "./ApproveModal";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof ApproveModal> = {
  title: "Components/ApproveModal",
  component: ApproveModal,
  tags: ["autodocs"],
  argTypes: {
    onApprove: { action: "onApprove" },
    onCancel: { action: "onCancel" },
  },
};

export default meta;

type Story = StoryObj<typeof ApproveModal>;

export const Shown: Story = {
  args: {
    title: "Sample Title",
    content: (
      <>
        This is a sample body text for the ApproveModal component.
        <br />
        <br />
        Please confirm your action.
      </>
    ),
    isOpen: true,
    cancelButtonText: "Cancel",
    approveButtonText: "Confirm",
  },
};
