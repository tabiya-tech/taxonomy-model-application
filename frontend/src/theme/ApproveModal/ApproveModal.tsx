import React from "react";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";
import CancelButton from "src/theme/CancelButton/CancelButton";
import { Dialog, DialogActions, DialogContent, DialogTitle } from "@mui/material";

interface ApproveModalProps {
  title: string;
  content: React.ReactElement;
  isOpen: boolean;
  onCancel: () => void;
  onApprove: () => void;
  cancelButtonText: string;
  approveButtonText: string;
}

const uniqueId = "edc0ac15-67a7-4c1f-9934-94fa65757046";

export const DATA_TEST_ID = {
  APPROVE_MODEL: `approve-model-${uniqueId}`,
  APPROVE_MODEL_TITLE: `approve-model-title-${uniqueId}`,
  APPROVE_MODEL_CONTENT: `approve-model-content-${uniqueId}`,
  APPROVE_MODEL_CANCEL: `approve-model-cancel-${uniqueId}`,
  APPROVE_MODEL_CONFIRM: `approve-model-confirm-${uniqueId}`,
};

const ApproveModal: React.FC<ApproveModalProps> = (props) => {
  return (
    <Dialog
      open={props.isOpen}
      onClose={props.onCancel}
      aria-labelledby="approve-model"
      data-testid={DATA_TEST_ID.APPROVE_MODEL}
      PaperProps={{
        sx: { padding: 4 },
      }}
    >
      <DialogTitle id="approve-model-title" data-testid={DATA_TEST_ID.APPROVE_MODEL_TITLE}>
        {props.title}
      </DialogTitle>
      <DialogContent data-testid={DATA_TEST_ID.APPROVE_MODEL_CONTENT} sx={{ marginTop: 4 }}>
        {props.content}
      </DialogContent>
      <DialogActions>
        <CancelButton onClick={props.onCancel} data-testid={DATA_TEST_ID.APPROVE_MODEL_CANCEL}>
          {props.cancelButtonText}
        </CancelButton>
        <PrimaryButton onClick={props.onApprove} data-testid={DATA_TEST_ID.APPROVE_MODEL_CONFIRM}>
          {props.approveButtonText}
        </PrimaryButton>
      </DialogActions>
    </Dialog>
  );
};

export default ApproveModal;
