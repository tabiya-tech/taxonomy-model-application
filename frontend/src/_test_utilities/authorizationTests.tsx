import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
