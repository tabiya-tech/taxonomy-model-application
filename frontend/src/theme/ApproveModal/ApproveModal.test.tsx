// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import ApproveModal, { DATA_TEST_ID } from "src/theme/ApproveModal/ApproveModal";

describe("ApproveModal", () => {
  test("should render component correctly", () => {
    // GIVEN the ApproveModal
    const givenNewConversationDialog = (
      <ApproveModal
        title="Sample Title?"
        content={
          <>
            This is a sample body text for the ApproveModal component.
            <br />
            <br />
            Please confirm your action.
          </>
        }
        isOpen={true}
        onCancel={() => {}}
        onApprove={() => {}}
        cancelButtonText="Cancel"
        approveButtonText="Confirm"
      />
    );

    // WHEN the ApproveModal is rendered
    render(givenNewConversationDialog);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND dialog container to be in the document
    const dialogContainer = screen.getByTestId(DATA_TEST_ID.APPROVE_MODEL);
    expect(dialogContainer).toBeInTheDocument();
    // AND dialog title to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.APPROVE_MODEL_TITLE)).toBeInTheDocument();
    // AND dialog content to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.APPROVE_MODEL_CONTENT)).toBeInTheDocument();
    // AND dialog cancel button to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.APPROVE_MODEL_CANCEL)).toBeInTheDocument();
    // AND dialog confirm button to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.APPROVE_MODEL_CONFIRM)).toBeInTheDocument();
    // AND dialog to match the snapshot
    expect(dialogContainer).toMatchSnapshot();
  });

  test("Action tests", () => {
    // GIVEN the cancel and approve functions
    const givenCancel = jest.fn();
    const givenApprove = jest.fn();

    // AND the ApproveModal
    const givenNewConversationDialog = (
      <ApproveModal
        title="Sample Title?"
        content={
          <>
            This is a sample body text for the ApproveModal component.
            <br />
            <br />
            Please confirm your action.
          </>
        }
        isOpen={true}
        onCancel={givenCancel}
        onApprove={givenApprove}
        cancelButtonText="Cancel"
        approveButtonText="Confirm"
      />
    );

    // WHEN the ApproveModal is rendered
    render(givenNewConversationDialog);

    // AND the cancel button is clicked
    screen.getByTestId(DATA_TEST_ID.APPROVE_MODEL_CANCEL).click();

    // THEN expect the cancel function to have been called
    expect(givenCancel).toHaveBeenCalled();

    // WHEN the confirm button is clicked
    screen.getByTestId(DATA_TEST_ID.APPROVE_MODEL_CONFIRM).click();

    // THEN expect the approval function to have been called
    expect(givenApprove).toHaveBeenCalled();
  });
});
