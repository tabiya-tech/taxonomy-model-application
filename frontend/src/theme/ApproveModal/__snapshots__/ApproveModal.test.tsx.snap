// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApproveModal should render component correctly 1`] = `
<div
  class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
  data-testid="approve-model-edc0ac15-67a7-4c1f-9934-94fa65757046"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    tabindex="-1"
  >
    <div
      aria-labelledby="approve-model"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1tsvg93-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 MuiDialogTitle-root css-pvqrlz-MuiTypography-root-MuiDialogTitle-root"
        data-testid="approve-model-title-edc0ac15-67a7-4c1f-9934-94fa65757046"
        id="approve-model-title"
      >
        Sample Title?
      </h2>
      <div
        class="MuiDialogContent-root css-1qs5egk-MuiDialogContent-root"
        data-testid="approve-model-content-edc0ac15-67a7-4c1f-9934-94fa65757046"
      >
        This is a sample body text for the ApproveModal component.
        <br />
        <br />
        Please confirm your action.
      </div>
      <div
        class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation css-140m69p-MuiButtonBase-root-MuiButton-root"
          data-testid="approve-model-cancel-edc0ac15-67a7-4c1f-9934-94fa65757046"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation css-fwypri-MuiButtonBase-root-MuiButton-root"
          data-testid="approve-model-confirm-edc0ac15-67a7-4c1f-9934-94fa65757046"
          tabindex="0"
          type="button"
        >
          Confirm
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
